rfcs/0042-config-option.md

```diff
@@ -0,0 +1,242 @@
+
+By not doing the tedious work of writing out dozens of options, module authors also have more time to do more meaningful work such as
+- Writing a NixOS test
+- Writing documentation
+- Implementing high-level options that tie different NixOS modules together in non-trivial ways (e.g. `enableACME`)
+
+Problem instances:
+- The [i2pd module](https://github.com/NixOS/nixpkgs/blob/2a669d3ee1308c7fd73f15beb35c0456ff9202bc/nixos/modules/services/networking/i2pd.nix) has a long [history](https://github.com/NixOS/nixpkgs/commits/2a669d3ee1308c7fd73f15beb35c0456ff9202bc/nixos/modules/services/networking/i2pd.nix) of option additions due to upstream updates, bug fixes and documentation changes
+- Because prometheus uses options to encode every possible setting, PR's like [#56017](https://github.com/NixOS/nixpkgs/pull/56017) are needed to allow users to set a part of the configuration that wasn't encoded yet.
+- Because strongswan-ctl uses options to encode its full configuration, changes like [#49197](https://github.com/NixOS/nixpkgs/pull/49197) are needed to update our options with upstream changes.
+
+These are only examples of where people *found* problems and fixed them. The number of modules that have outdated options and require maintenance is probably much higher.
+
+
+# Detailed design
+[design]: #detailed-design
+
+## [Part 1][part1]
+[part1-design]: #part-1-1
+
+It's already possible to write generic `settings` options today, using PR's introducing [`pkgs.formats`](https://github.com/NixOS/nixpkgs/pull/75584) and [freeform modules](https://github.com/NixOS/nixpkgs/pull/82743) when needed. Documentation for these features and how to use them for declaring `settings` options already exists in the manual: [Options for Program Settings](https://nixos.org/manual/nixos/stable/index.html#sec-settings-options) and [Freeform Modules](https://nixos.org/manual/nixos/stable/index.html#sec-freeform-modules).
+
+Using these features, a module supporting `settings` might look like
+
+```nix
+{ options, config, lib, pkgs, ... }:
+let
+  cfg = config.services.foo;
+  # Define the settings format used for this program
+  settingsFormat = pkgs.formats.json {};
+in {
+
+  options.services.foo = {
+    enable = lib.mkEnableOption "foo service";
+
+    settings = lib.mkOption {
+      type = lib.types.submodule {
+
+        # Declare that the settings option supports arbitrary format values, json here
+        freeformType = settingsFormat.type;
+
+        # Declare an option for the port such that the type is checked and this option
+        # is shown in the manual.
+        options.port = lib.mkOption {
+          type = lib.types.port;
+          default = 8080;
+          description = ''
+            Which port this service should listen on.
+          '';
+        };
+
+      };
+      default = {};
```

Comment:
:warning:  This should always be empty. All defaults that are specified here are forgotten when a configuration sets a setting.

This happens because specifying defaults has the effect of `lib.mkOptionDefault` (like `lib.mkDefault`) on the _whole_ attribute set instead of individual settings.

Likewise, you could make the same mistake in the `config.x.y.z.settings` definition, assuming `{ a = lib.mkDefault 1; }` would be equivalent to `lib.mkDefault { a = 1; }`, which it is not. Only the prior preserves `a` when another setting (not `a`) is defined by the user.

For a real life fix, see https://github.com/NixOS/nixpkgs/pull/110614

@Infinisil, perhaps we could make `default = {}` implied whenever `freeformType` is set? It'd be less tempting.

Reply:
@roberth That reads to me like an important use case to think through thoroughly in `nickel` (in case it's not already a in depth first tree merge)

Reply:
@blaggacao then please take responsibility and post on that repo instead. Let's keep this PR on-topic, respecting the 40+ participants' time and attention. Thank you.

Reply:
I'm not an octopus, somteimes just linear associations, that I dump. sorry.

Reply:
I've been planning to make a PR for allowing something like
```nix
mkOption {
  recursiveDefault = {
    foo.bar = 10;
  };
}
```

Which would be equivalent to
```nix
{
  config.foo.bar = mkOptionDefault 10;
}
```

This would also allow rendering the default in the manual